Intermittent Cirrus test failure: test_fetch_schedule_recipes_retry

Seen here: https://app.circleci.com/pipelines/github/mozilla/experimenter/59098/workflows/4129ed68-c286-41f5-8f9f-a34a5c84e6b8/jobs/347195

```py
______________________ test_fetch_schedule_recipes_retry _______________________

self = <MagicMock name='scheduler.add_job' id='137473121149072'>
args = (<function fetch_schedule_recipes at 0x7d07f4affce0>, 'interval')
kwargs = {'max_instances': 1, 'max_retries': 3, 'seconds': 30}
msg = "Expected 'add_job' to be called once. Called 2 times.\nCalls: [call(<function fetch_schedule_recipes at 0x7d07f4affce...\n call(<function fetch_schedule_recipes at 0x7d07f4affce0>, 'interval', seconds=30, max_instances=1, max_retries=3)]."

    def assert_called_once_with(self, /, *args, **kwargs):
        """assert that the mock was called exactly once and that that call was
        with the specified arguments."""
        if not self.call_count == 1:
            msg = ("Expected '%s' to be called once. Called %s times.%s"
                   % (self._mock_name or 'mock',
                      self.call_count,
                      self._calls_repr()))
>           raise AssertionError(msg)
E           AssertionError: Expected 'add_job' to be called once. Called 2 times.
E           Calls: [call(<function fetch_schedule_recipes at 0x7d07f4affce0>, 'interval', seconds=30, max_instances=1, max_retries=3),
E            call(<function fetch_schedule_recipes at 0x7d07f4affce0>, 'interval', seconds=30, max_instances=1, max_retries=3)].

/usr/local/lib/python3.11/unittest/mock.py:950: AssertionError

During handling of the above exception, another exception occurred:

scheduler_mock = <MagicMock name='scheduler' id='137473121146192'>
remote_setting_live_mock = <MagicMock name='remote_setting_live' id='137473121978320'>
exception = Exception('some error')

    @pytest.mark.asyncio
    async def test_fetch_schedule_recipes_retry(
        scheduler_mock, remote_setting_live_mock, exception
    ):
        # Set up the remote_setting_mock to raise an exception the first time it is called
        # and return a value the second time it is called
        remote_setting_live_mock.fetch_recipes.side_effect = [
            exception,
            ["recipe1", "recipe2"],
        ]
    
        await fetch_schedule_recipes()
    
        # Check that a job was added to the scheduler to retry after 30 seconds
>       scheduler_mock.add_job.assert_called_once_with(
            fetch_schedule_recipes,
            "interval",
            seconds=30,
            max_instances=1,
            max_retries=3,
        )
E       AssertionError: Expected 'add_job' to be called once. Called 2 times.
E       Calls: [call(<function fetch_schedule_recipes at 0x7d07f4affce0>, 'interval', seconds=30, max_instances=1, max_retries=3),
E        call(<function fetch_schedule_recipes at 0x7d07f4affce0>, 'interval', seconds=30, max_instances=1, max_retries=3)].

tests/test_main.py:407: AssertionError
------------------------------ Captured log setup ------------------------------
```

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-5766)

--- a/cirrus/server/tests/conftest.py
+++ b/cirrus/server/tests/conftest.py
@@ -40,13 +40,19 @@ def client():
 
 
 @fixture
-def scheduler_mock():
+def app_state_mock():
+    with mock.patch("cirrus.main.app.state") as app_state_mock:
+        yield app_state_mock
+
+
+@fixture
+def scheduler_mock(app_state_mock):
     with mock.patch("cirrus.main.app.state.scheduler") as scheduler_mock:
         yield scheduler_mock
 
 
 @fixture
-def remote_setting_live_mock():
+def remote_setting_live_mock(app_state_mock):
     with mock.patch(
         "cirrus.main.app.state.remote_setting_live"
     ) as remote_setting_live_mock:
@@ -54,7 +60,7 @@ def remote_setting_live_mock():
 
 
 @fixture
-def remote_setting_preview_mock():
+def remote_setting_preview_mock(app_state_mock):
     with mock.patch(
         "cirrus.main.app.state.remote_setting_preview"
     ) as remote_setting_preview_mock:
@@ -106,11 +112,6 @@ def fml_with_coenrolling_features_path():
 @fixture
 def fml_with_coenrolling_features(fml_with_coenrolling_features_path):
     return FeatureManifestLanguage(fml_with_coenrolling_features_path, channel)
-
-
-@fixture
-def exception():
-    return Exception("some error")
 
 
 @fixture(name="reset_glean", scope="function", autouse=True)

--- a/cirrus/server/tests/test_main.py
+++ b/cirrus/server/tests/test_main.py
@@ -247,170 +247,87 @@ def test_get_features_v2_missing_required_field(
 
 
 @pytest.mark.asyncio
-@patch("cirrus.main.app.state.remote_setting_live")
-@patch("cirrus.main.app.state.remote_setting_preview")
-async def test_fetch_schedule_recipes_success(
-    mock_remote_setting_live, mock_remote_setting_preview, scheduler_mock
+@pytest.mark.parametrize(
+    "live_side_effect, preview_side_effect, has_preview, expect_retry",
+    [
+        pytest.param(
+            None,
+            None,
+            False,
+            False,
+            id="live_succeeds_no_preview",
+        ),
+        pytest.param(
+            None,
+            None,
+            True,
+            False,
+            id="both_succeed",
+        ),
+        pytest.param(
+            Exception("live error"),
+            None,
+            True,
+            True,
+            id="live_fails_no_preview",
+        ),
+        pytest.param(
+            Exception("live error"),
+            None,
+            True,
+            True,
+            id="live_fails_preview_succeeds",
+        ),
+        pytest.param(
+            None,
+            Exception("preview error"),
+            True,
+            True,
+            id="live_succeeds_preview_fails",
+        ),
+        pytest.param(
+            Exception("live error"),
+            Exception("preview error"),
+            True,
+            True,
+            id="both_fail",
+        ),
+    ],
+)
+async def test_fetch_schedule_recipes(
+    remote_setting_live_mock,
+    remote_setting_preview_mock,
+    scheduler_mock,
+    live_side_effect,
+    preview_side_effect,
+    has_preview,
+    expect_retry,
 ):
-    mock_remote_setting_live.fetch_recipes.return_value = {
-        "data": [{"experiment1": True}, {"experiment2": False}]
-    }
-    mock_remote_setting_preview.fetch_recipes.return_value = {
-        "data": [{"experiment1": True}, {"experiment2": False}]
-    }
-
-    await fetch_schedule_recipes()
-    mock_remote_setting_live.fetch_recipes.assert_called_once()
-    mock_remote_setting_preview.fetch_recipes.assert_called_once()
-
-    # Check that no jobs were added to the scheduler
-    scheduler_mock.add_job.assert_not_called()
-
-
-@pytest.mark.asyncio
-@patch("cirrus.main.app.state.scheduler")
-@patch("cirrus.main.app.state.remote_setting_live")
-async def test_fetch_schedule_recipes_failure_live(
-    mock_remote_setting_live, scheduler_mock
-):
-    mock_remote_setting_live.fetch_recipes.side_effect = Exception("some error")
-
-    await fetch_schedule_recipes()
-
-    mock_remote_setting_live.fetch_recipes.assert_called_once()
-
-    # Check that a job was added to the scheduler to retry after 30 seconds
-    scheduler_mock.add_job.assert_called_once_with(
-        fetch_schedule_recipes,
-        "interval",
-        seconds=30,
-        max_instances=1,
-        max_retries=3,
-    )
-
-
-@pytest.mark.asyncio
-@patch("cirrus.main.app.state.scheduler")
-@patch("cirrus.main.app.state.remote_setting_preview")
-async def test_fetch_schedule_recipes_failure_preview(
-    mock_remote_setting_preview, scheduler_mock
-):
-    mock_remote_setting_preview.fetch_recipes.side_effect = Exception("some error")
-
-    await fetch_schedule_recipes()
-
-    mock_remote_setting_preview.fetch_recipes.assert_called_once()
-
-    # Check that a job was added to the scheduler to retry after 30 seconds
-    scheduler_mock.add_job.assert_called_once_with(
-        fetch_schedule_recipes,
-        "interval",
-        seconds=30,
-        max_instances=1,
-        max_retries=3,
-    )
-
-
-@pytest.mark.asyncio
-@patch("cirrus.main.app.state.scheduler")
-@patch("cirrus.main.app.state.remote_setting_live")
-@patch("cirrus.main.app.state.remote_setting_preview")
-async def test_fetch_schedule_recipes_both_fail(
-    mock_remote_setting_live, mock_remote_setting_preview, scheduler_mock
-):
-    mock_remote_setting_live.fetch_recipes.side_effect = Exception("live error")
-    mock_remote_setting_preview.fetch_recipes.side_effect = Exception("preview error")
+    if live_side_effect is not None:
+        remote_setting_live_mock.fetch_recipes.side_effect = live_side_effect
+    if preview_side_effect is not None:
+        remote_setting_preview_mock.fetch_recipes.side_effect = preview_side_effect
+    remote_setting_preview_mock.__bool__.return_value = has_preview
 
     await fetch_schedule_recipes()
 
-    mock_remote_setting_live.fetch_recipes.assert_called_once()
-    mock_remote_setting_preview.fetch_recipes.assert_called_once()
+    remote_setting_live_mock.fetch_recipes.assert_called_once()
 
-    # Check that only one job was added to the scheduler to retry after 30 seconds
-    scheduler_mock.add_job.assert_called_once_with(
-        fetch_schedule_recipes,
-        "interval",
-        seconds=30,
-        max_instances=1,
-        max_retries=3,
-    )
+    if has_preview:
+        remote_setting_preview_mock.fetch_recipes.assert_called_once()
+    else:
+        remote_setting_preview_mock.fetch_recipes.assert_not_called()
 
-
-@pytest.mark.asyncio
-@patch("cirrus.main.app.state.scheduler")
-@patch("cirrus.main.app.state.remote_setting_live")
-@patch("cirrus.main.app.state.remote_setting_preview")
-async def test_fetch_schedule_recipes_live_fails_preview_succeeds(
-    mock_remote_setting_live, mock_remote_setting_preview, scheduler_mock
-):
-    mock_remote_setting_live.fetch_recipes.side_effect = Exception("live error")
-    mock_remote_setting_preview.fetch_recipes.return_value = {
-        "data": [{"experiment1": True}, {"experiment2": False}]
-    }
-
-    await fetch_schedule_recipes()
-
-    mock_remote_setting_live.fetch_recipes.assert_called_once()
-    mock_remote_setting_preview.fetch_recipes.assert_called_once()
-
-    # Check that a job was added to the scheduler to retry after 30 seconds
-    scheduler_mock.add_job.assert_called_once_with(
-        fetch_schedule_recipes,
-        "interval",
-        seconds=30,
-        max_instances=1,
-        max_retries=3,
-    )
-
-
-@pytest.mark.asyncio
-@patch("cirrus.main.app.state.scheduler")
-@patch("cirrus.main.app.state.remote_setting_live")
-@patch("cirrus.main.app.state.remote_setting_preview")
-async def test_fetch_schedule_recipes_preview_fails_live_succeeds(
-    mock_remote_setting_live, mock_remote_setting_preview, scheduler_mock
-):
-    mock_remote_setting_live.fetch_recipes.return_value = {
-        "data": [{"experiment1": True}, {"experiment2": False}]
-    }
-    mock_remote_setting_preview.fetch_recipes.side_effect = Exception("preview error")
-
-    await fetch_schedule_recipes()
-
-    mock_remote_setting_live.fetch_recipes.assert_called_once()
-    mock_remote_setting_preview.fetch_recipes.assert_called_once()
-
-    # Check that a job was added to the scheduler to retry after 30 seconds
-    scheduler_mock.add_job.assert_called_once_with(
-        fetch_schedule_recipes,
-        "interval",
-        seconds=30,
-        max_instances=1,
-        max_retries=3,
-    )
-
-
-@pytest.mark.asyncio
-async def test_fetch_schedule_recipes_retry(
-    scheduler_mock, remote_setting_live_mock, exception
-):
-    # Set up the remote_setting_mock to raise an exception the first time it is called
-    # and return a value the second time it is called
-    remote_setting_live_mock.fetch_recipes.side_effect = [
-        exception,
-        ["recipe1", "recipe2"],
-    ]
-
-    await fetch_schedule_recipes()
-
-    # Check that a job was added to the scheduler to retry after 30 seconds
-    scheduler_mock.add_job.assert_called_once_with(
-        fetch_schedule_recipes,
-        "interval",
-        seconds=30,
-        max_instances=1,
-        max_retries=3,
-    )
+    if expect_retry:
+        scheduler_mock.add_job.assert_called_once_with(
+            fetch_schedule_recipes,
+            "interval",
+            seconds=30,
+            max_instances=1,
+            max_retries=3,
+        )
+    else:
+        scheduler_mock.add_job.assert_not_called()
 
 
 def test_lbheartbeat_endpoint(client):

